crates/wasmparser/src/readers/core/reloc.rs

```diff
@@ -0,0 +1,307 @@
+use std::ops::Range;
+
+use crate::{BinaryReader, FromReader, Result, SectionLimited};
+
+/// Reader for relocation entries within a `reloc.*` section.
+pub type RelocationEntryReader<'a> = SectionLimited<'a, RelocationEntry>;
+
+/// Reader for reloc.* sections as defined by
+/// https://github.com/WebAssembly/tool-conventions/blob/main/Linking.md#relocation-sections.
+#[derive(Debug, Clone)]
+pub struct RelocSectionReader<'a> {
+    section: u32,
+    range: Range<usize>,
+    entries: SectionLimited<'a, RelocationEntry>,
+}
+
+impl<'a> RelocSectionReader<'a> {
+    /// Creates a new reader for a `reloc.*` section starting at
+    /// `original_position` within the wasm file.
+    pub fn new(data: &'a [u8], original_position: usize) -> Result<Self> {
+        Self::from_reader(&mut BinaryReader::new_with_offset(data, original_position))
+    }
+
+    /// Index of section to which the relocations apply.
+    pub fn section_index(&self) -> u32 {
+        self.section
+    }
+
+    /// The byte range of the entire section.
+    pub fn range(&self) -> Range<usize> {
+        self.range.clone()
+    }
+
+    /// The relocation entries.
+    pub fn entries(&self) -> SectionLimited<'a, RelocationEntry> {
+        self.entries.clone()
+    }
+}
+
+impl<'a> FromReader<'a> for RelocSectionReader<'a> {
+    fn from_reader(reader: &mut BinaryReader<'a>) -> Result<Self> {
+        let range = reader.range().clone();
+        let section = reader.read_var_u32()?;
+        Ok(Self {
+            section,
+            range,
+            entries: SectionLimited::new(reader.remaining_buffer(), reader.original_position())?,
+        })
+    }
+}
```

Comment:
Mind dropping this impl? The `FromReader` trait is mostly intended for "parse the next few bytes as this type" but this implementation is taking the whole `BinaryReader` contents, so I think it'd be best to package this up in the `new` constructor above

Reply:
Done.